### PR TITLE
rfc: propose pow and rsqrt op for graph api

### PR DESCRIPTION
**Pull request by rongzha1**
Pow and Rsqrt are requested to add in Graph API for LLM optimization to get a better performance..

This rfc provides 5 options to support these 2 OPs.

Co-work with @wzt1997

### CODE REVIEW

**Comment by mgouicem** [comment]
Hi @rongzha1 and thanks for the RFC. I guess there are two things here:
1. expose pow/rsqrt as separate ops or as the same op
2. expose exponent as attribute or input tensor.

For 1. I would recommend to expose only pow. The reason for that is that the backend will have to optimize the path for exponent=-1/2 anyway. Also, fewer op means simpler validation and more robust optimization.

For 2, do you have examples of models or frameworks where the exponent is not a constant and comes as a non constant tensor? I have a hard time thinking of use-cases for exponent coming from previous layer (high likelihood of producing underflows and overflows). I think we really need to understand this part clearly before making a decision, as it has deep implication in backend implementation and performance.

**Comment by rongzha1** [comment]
> 1. expose pow/rsqrt as separate ops or as the same op
> 2. expose exponent as attribute or input tensor.
> 
> For 1. I would recommend to expose only pow. The reason for that is that the backend will have to optimize the path for exponent=-1/2 anyway. Also, fewer op means simpler validation and more robust optimization.
> 
> For 2, do you have examples of models or frameworks where the exponent is not a constant and comes as a non constant tensor? I have a hard time thinking of use-cases for exponent coming from previous layer (high likelihood of producing underflows and overflows). I think we really need to understand this part clearly before making a decision, as it has deep implication in backend implementation and performance.

Thanks for your suggestion. 
For 1.  `rsqrt` map to `pow` with attribute -1/2 seems easy for frameworks,
            but map to `pow` with two input tensor ipex colleague has some concerns:
            1. wrap -1/2 to tensor may change graph 
            2. lost the constant info after -1/2 changed to a tensor.
So the question can be can be simplified to 
    1. expose attribute pow, which can cover rsqrt 
    2. how to support binary power, is it necessary or a must to support this cfg?
Most frameworks and toolkit takes exponent as second input tensor. But in this case, exponent is cosnt scalar 2.  @chunyuan-w @retonym Could you provide some example runtime exponent coming from previous layer?

File: rfcs/20230418-add-pow-graph-op/README.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+popular in the open-source community, which requires oneDNN to provide
+corresponding support and optimization.
+
+`Pow` and `Rsqrt` are used in the Pytorch version of LLaMa and GPT-J. The
+framework team is requesting oneDNN Graph API to support the two operations and
+the graph compiler backend to fuse and optimize them.
+
+### Investigation
+
+`Pow` and `Rsqrt` are used to constructy RMSNorm, which is a variant of L2
```
Comment (ZhennanQin) [inline rfcs/20230418-add-pow-graph-op/README.md:17]: constructy ->　construct

Comment (rongzha1) [reply rfcs/20230418-add-pow-graph-op/README.md]: fixed, thanks

File: rfcs/20230418-add-pow-graph-op/README.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+Pros.
+
+1. `Pow` is compatible with mainstream frameworks.
+2. `Rsqrt` reuse current op set. Will keep the op set small.
+
+Cons.
+
+1. framework bridge will need to split `Rsqrt` to two oneDNN Graph operations.
+2. `Pow` cannot be optimized with oneDNN primitive.
```
Comment (retonym) [inline rfcs/20230418-add-pow-graph-op/README.md:232]: why `Pow` cannot be optimized with oneDNN here

Comment (rongzha1) [reply rfcs/20230418-add-pow-graph-op/README.md:232]: in this option, the second input is a runtime tensor. but  oneDNN only support exponent as single constant attribute
https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html

Comment (retonym) [reply rfcs/20230418-add-pow-graph-op/README.md:232]: Thanks for the explanation.

**Comment by rongzha1** [comment]
update the rfc with meeting minutes, option 1 is chosed as current solution:
unary eltwise Pow with attribute, Rsqrt reused Pow with attribute -1/2

1. Easy for framework to integrate
2. Can leverage oneDNN primitive
3. Don’t have runtime exponent Pow in current models

For runtime exponent, need revisit this proposal in the future. Maybe add a new
operation dynamicPow.

could you review it again?

**Review by ZhennanQin — APPROVED** [review]
LGTM.

File: rfcs/20230418-add-pow-graph-op/README.md [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+Add eltwise `Pow` that takes a single input tensor src, with exponent as an
+single-value attribute.
+`Rsqrt` reuses `Pow` with exponent attribute -1/2.
+
+| Pow       | Argument Name    | Required or Optional       | Data Type     |
+|-----------|------------------|----------------------------|---------------|
+| input     | `src`            | Required                   | f32,f16,bf16* |
+| attribute | `exponent`       | Required                   | f32           |
+| attribute | `auto_broadcast` | optional, default: 'numpy' | string        |
```
Comment (TaoLv) [inline rfcs/20230418-add-pow-graph-op/README.md:101]: For this option, what do you mean by supporting broadcasting?

Comment (rongzha1) [reply rfcs/20230418-add-pow-graph-op/README.md]: Good catch. Thanks you. `auto_broadcast` only works for binary pow. fixed

File: rfcs/20230418-add-pow-graph-op/README.md [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+`exponent` is optional as well. It accepts two configurations: eltwise Pow with
+attribute `exponent` and binary Pow without attribute `exponent`.
+`Rsqrt` reuses `Pow` with exponent attribute -1/2.
+
+| Pow       | Argument Name    | Required or Optional        | Data Type     |
+|-----------|------------------|-----------------------------|---------------|
+| input     | `src`            | Required                    | f32,f16,bf16* |
+| input     | `exponent`       | optional*                   | f32,f16,bf16  |
+| attribute | `auto_broadcast` | optional, default: 'numpy'  | f32           |
+| attribute | `exponent`       | optional*                   | string        |
```
Comment (wuxun-zhang) [inline rfcs/20230418-add-pow-graph-op/README.md:131]:
```suggestion
| attribute | `auto_broadcast` | optional, default: 'numpy'  | string        |
| attribute | `exponent`       | optional*                   | f32           |
```

Comment (rongzha1) [reply rfcs/20230418-add-pow-graph-op/README.md]: fixed. Thank you

**Comment by rongzha1** [comment]
doc has been updated, coud you review it again @mgouicem @wuxun-zhang @retonym ?

File: rfcs/20230418-add-pow-graph-op/README.md [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+| output    | dst         | dst     |       |           |         |
+
+*：ONNX, OpenVINO and oneDNN does not support Rsqrt directly, but in ONNX and
+OpenVINO Rsqrt can be supported by Pow(src, -1/2) and in oneDNN it is supported
+by Pow(src) with attribute -1/2.
+
+## Proposal
+
+To support the request from framework, we need to add new operations to graph
+API, with 5 options provides as follows:
```
Comment (chunyuan-w) [inline rfcs/20230418-add-pow-graph-op/README.md:89]: nit: `provides` -> `provided`

**Review by chunyuan-w — APPROVED** [review]

File: rfcs/20230418-add-pow-graph-op/README.md (hunk 5 above)
Comment (rongzha1) [reply rfcs/20230418-add-pow-graph-op/README.md]: fixed. thanks

**Review by retonym — APPROVED** [review]

File: rfcs/20230418-add-pow-graph-op/README.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+For runtime exponent, need revisit this proposal in the future. Maybe add a new
+operation dynamicPow.
+### API changes
+
+New operation(s) will be added in Graph API.
+
+### Backend changes
+
+Graph Compiler backend will fuse new added operation(s) into MLP pattern for LLM
```
Comment (igorsafo) [inline rfcs/20230418-add-pow-graph-op/README.md:251]: @TaoLv @rongzha1 Will primitive-based backend support this op? Eltwise primitive does support pow algorithm.

**Review by igorsafo — APPROVED** [review]

File: rfcs/20230418-add-pow-graph-op/README.md (hunk 6 above)
Comment (rongzha1) [reply rfcs/20230418-add-pow-graph-op/README.md:251]: primitive-based backend will be added on demand. Currently no single Pow optimization request from frameworks.
Or can we target it on v3.3?

Comment (TaoLv) [reply rfcs/20230418-add-pow-graph-op/README.md:251]: Yes, once the op is added, we will support the post-ops fusion based on the new op as usual, eg. conv/matmul + pow patterns. But so far we don't have the plan to support the RMSNorm fusion mentioned here.

**Review by TaoLv — APPROVED** [review]
LGTM

**Comment by TaoLv** [comment]
The RFC has been implementation on master branch via d97c2d2e6d5b2b9392ae0673e4e55b8c1515b9e9, 84242fb9d13759d1c16ec83ac7593170b833c710, and 58c40d03c13ed612659303919c2a73e20ebc82ac .

**Pull request merged**